File: Components/DevExpressVCL/Demos/ExpressQuantumTreeList/CBuilder/DragDropDemo/DragDropDemoDictionary.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "DragDropDemoDictionary.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxCheckBox"
#pragma link "cxControls"
#pragma link "cxCurrencyEdit"
#pragma link "cxCustomData"
#pragma link "cxDBTL"
#pragma link "cxGraphics"
#pragma link "cxInplaceContainer"
#pragma link "cxMaskEdit"
#pragma link "cxStyles"
#pragma link "cxTL"
#pragma link "cxTLData"
#pragma resource "*.dfm"
TDragDropDemoDictionaryForm *DragDropDemoDictionaryForm;

int __fastcall cxCompareNodes(void *AItem1, void *AItem2)
{
  return ((TcxDBTreeListNode*)AItem2)->KeyValue -
    ((TcxDBTreeListNode*)AItem1)->KeyValue;
}

void SetSelectedNodesValue(TcxDBTreeList *ATreeList, int AItemIndex,
  Variant AValue)
{
  TList *AList = new TList();
  ATreeList->BeginUpdate();
  __try{
    ATreeList->GetSelections(AList);
    AList->Sort(cxCompareNodes);
    for (int I = 0; I < AList->Count; I++)
      if (((TcxTreeListNode*)AList->Items[I])->Values[AItemIndex] != AValue){
        ((TcxTreeListNode*)AList->Items[I])->Focused = true;
        ATreeList->DataController->Edit();
        ((TcxTreeListNode*)AList->Items[I])->Values[AItemIndex] = AValue;
        ATreeList->DataController->Post();
      }
    }
  __finally{
    delete AList;
    ATreeList->EndUpdate();
  }
}

//---------------------------------------------------------------------------

__fastcall TDragDropDemoDictionaryForm::TDragDropDemoDictionaryForm(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TDragDropDemoDictionaryForm::tlDictDragOver(TObject *Sender,
  TObject* Source, int X, int Y, TDragState State, bool &Accept)
{
  Accept = false;
}
//---------------------------------------------------------------------------

void __fastcall TDragDropDemoDictionaryForm::tlDictMoveTo(TcxCustomTreeList *Sender,
  TcxTreeListNode *AttachNode, TcxTreeListNodeAttachMode AttachMode,
  TList *Nodes, bool &IsCopy, bool &Done)
{
  Done = true;
}
//---------------------------------------------------------------------------

void TDragDropDemoDictionaryForm::SetDeptSelectionParentValue(Variant AParentValue)
{
  SetSelectedNodesValue(tlDeptDict, tlDeptDictPARENTID->ItemIndex, AParentValue);
}
//---------------------------------------------------------------------------

void TDragDropDemoDictionaryForm::SetEmplSelectionDeptID(Variant ADepartmentID)
{
  SetSelectedNodesValue(tlEmplDict, tlEmplDictDepartmentID->ItemIndex, ADepartmentID);
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressQuantumTreeList/Delphi/DragDropDemo/DragDropDemoDictionary.pas
unit DragDropDemoDictionary;

{$I cxVer.inc}

interface

uses
  Windows, Messages, SysUtils, Variants, Classes,
  Graphics, Controls, Forms, Dialogs, ComCtrls, StdCtrls, DragDropDemoData,
  cxGraphics, cxCustomData, cxStyles, cxTL, cxControls, cxInplaceContainer,
  cxTLData, cxDBTL, cxMaskEdit, cxCurrencyEdit, cxMemo, cxCheckBox;

type
  TDragDropDemoDictionaryForm = class(TForm)
    pcDictionary: TPageControl;
    tsDepartments: TTabSheet;
    tsPersons: TTabSheet;
    tlDeptDict: TcxDBTreeList;
    tlEmplDict: TcxDBTreeList;
    lsc: TLabel;
    Label1: TLabel;
    tlDeptDictPARENTID: TcxDBTreeListColumn;
    tlDeptDictNAME: TcxDBTreeListColumn;
    tlDeptDictBUDGET: TcxDBTreeListColumn;
    tlDeptDictPHONE: TcxDBTreeListColumn;
    tlDeptDictFAX: TcxDBTreeListColumn;
    tlDeptDictEMAIL: TcxDBTreeListColumn;
    tlDeptDictVACANCY: TcxDBTreeListColumn;
    tlEmplDictName: TcxDBTreeListColumn;
    tlEmplDictCountry: TcxDBTreeListColumn;
    tlEmplDictPostalCode: TcxDBTreeListColumn;
    tlEmplDictCity: TcxDBTreeListColumn;
    tlEmplDictAddress: TcxDBTreeListColumn;
    tlEmplDictPhone: TcxDBTreeListColumn;
    tlEmplDictFax: TcxDBTreeListColumn;
    tlEmplDictEMAIL: TcxDBTreeListColumn;
    tlEmplDictHOMEPAGE: TcxDBTreeListColumn;
    tlEmplDictDepartmentID: TcxDBTreeListColumn;
    procedure tlDictDragOver(Sender, Source: TObject; X, Y: Integer;
      State: TDragState; var Accept: Boolean);
    procedure tlDictMoveTo(Sender: TcxCustomTreeList;
      AttachNode: TcxTreeListNode; AttachMode: TcxTreeListNodeAttachMode;
      Nodes: TList; var IsCopy, Done: Boolean);
  public
    procedure SetDeptSelectionParentValue(AParentValue: Variant);
    procedure SetEmplSelectionDeptID(ADepartmentID: Variant);
  end;

procedure SetSelectedNodesValue(ATreeList: TcxDBTreeList; AItemIndex: Integer;
  AValue: Variant);

var
  DragDropDemoDictionaryForm: TDragDropDemoDictionaryForm;

implementation

{$R *.dfm}
function cxCompareNodes(AItem1, AItem2: Pointer): Integer;
begin
  with TcxDBTreeListNode(AItem1) do
    Result := TcxDBTreeListNode(AItem2).KeyValue - KeyValue;
end;

procedure SetSelectedNodesValue(ATreeList: TcxDBTreeList; AItemIndex: Integer; AValue: Variant);
var
  I: Integer;
  AList: TList;
begin
  with ATreeList do
  begin
    AList := TList.Create;
    BeginUpdate;
    try
      ATreeList.GetSelections(AList);
      AList.Sort(cxCompareNodes);
      for I := 0 to AList.Count - 1 do
        if TcxTreeListNode(AList[I]).Values[AItemIndex] <> AValue then
        begin

          TcxTreeListNode(AList[I]).Focused := True;
          DataController.Edit;
          TcxTreeListNode(AList[I]).Values[AItemIndex] := AValue;
          DataController.Post;
        end;
    finally
      AList.Free;
      EndUpdate;
    end;
  end
end;

procedure TDragDropDemoDictionaryForm.SetDeptSelectionParentValue(
  AParentValue: Variant);
begin
  SetSelectedNodesValue(tlDeptDict, tlDeptDictPARENTID.ItemIndex, AParentValue);
end;

procedure TDragDropDemoDictionaryForm.SetEmplSelectionDeptID(
  ADepartmentID: Variant);
begin
  SetSelectedNodesValue(tlEmplDict, tlEmplDictDepartmentID.ItemIndex, ADepartmentID);
end;

procedure TDragDropDemoDictionaryForm.tlDictDragOver(Sender,
  Source: TObject; X, Y: Integer; State: TDragState; var Accept: Boolean);
begin
  Accept := False;
end;

procedure TDragDropDemoDictionaryForm.tlDictMoveTo(Sender: TcxCustomTreeList;
  AttachNode: TcxTreeListNode; AttachMode: TcxTreeListNodeAttachMode;
  Nodes: TList; var IsCopy, Done: Boolean);
begin
  Done := True;
end;

end.

